receiver/factory.go

```diff
@@ -87,7 +87,11 @@ type FactoryV2 interface {
 	CreateTraceReceiverV2(ctx context.Context, logger *zap.Logger, cfg configmodels.Receiver,
 		nextConsumer consumer.TraceConsumerV2) (TraceReceiver, error)
 
-	// TODO: add CreateMetricsReceiverV2.
+	// CreateMetricReceiverV2 creates a metric receiver based on this config.
+	// If the receiver type does not support metrics or if the config is not valid
+	// error will be returned instead.
+	CreateMetricReceiverV2(ctx context.Context, logger *zap.Logger, cfg configmodels.Receiver,
```

Comment:
You may have already considered this but if the only change is adding context could these be named something like CreateMetricReceiverWithContext instead. Or, if we think more parameters may be passed in the future but we don't want to break backwards compatibility should we pass a struct where new members can be added but won't require changing function signature?

Reply:
I like the idea but I think it is inevitable that will have to break this again eventually because of the logger... we could try to not pass the logger here but that is going to be a much bigger change.

Reply:
Using a struct is a possibility. It was also suggested e.g. in this PR for application creation: https://github.com/open-telemetry/opentelemetry-collector/pull/621/files

I do like the idea a bit, but not totally sold on it, since it looks a bit unconventional. So, let's say we do this, what's your opinion, do you like it:

```go
type CreationParams struct {
	ctx    context.Context
	logger *zap.Logger
}

// FactoryV2 can create TraceExporterV2 and MetricsExporterV2. This is the
// new factory type that can create new style exporters.
type FactoryV2 interface {
	BaseFactory

	// CreateTraceReceiverV2 creates a trace receiver based on this config.
	// If the receiver type does not support tracing or if the config is not valid
	// error will be returned instead.
	CreateTraceExporterV2(params CreationParams, cfg configmodels.Exporter) (TraceExporterV2, error)

	// CreateMetricsExporterV2 creates a metrics receiver based on this config.
	// If the receiver type does not support metrics or if the config is not valid
	// error will be returned instead.
	CreateMetricsExporterV2(params CreationParams, cfg configmodels.Exporter) (MetricsExporterV2, error)
}

// Usage
params := CreationParams{ctx:context.Background(), logger: app.logger}
exporter := factory.CreateTraceExporterV2(params, config)
```

@bogdandrutu what do you think?

Reply:
I briefly looked and later changing the logger and it won't be so bad as I feared because many don't use it. So passing as parameter struct seems a good deal even now: it won't prevent all future break but reduce it only to the users of it. Side note: the standard practice for context is to keep it as the first parameter of the function and the logger field needs to be public.

Reply:
@pjanotti So you are suggesting this:
```go
type CreationParams struct {
	Logger *zap.Logger
}

// FactoryV2 can create TraceExporterV2 and MetricsExporterV2. This is the
// new factory type that can create new style exporters.
type FactoryV2 interface {
	BaseFactory

	// CreateTraceReceiverV2 creates a trace receiver based on this config.
	// If the receiver type does not support tracing or if the config is not valid
	// error will be returned instead.
	CreateTraceExporterV2(ctx context.Context, params CreationParams, cfg configmodels.Exporter) (TraceExporterV2, error)

	// CreateMetricsExporterV2 creates a metrics receiver based on this config.
	// If the receiver type does not support metrics or if the config is not valid
	// error will be returned instead.
	CreateMetricsExporterV2(ctx context.Context, params CreationParams, cfg configmodels.Exporter) (MetricsExporterV2, error)
}
```

I am not sure I understand your concern with logger.

Reply:
Yes, that my suggestion. I'm assuming that some time down the road we will have an OpenTelemetry logger and initially I thought that would be a lot of work when that was changed but is less used than what I thought - so let's go for it :)

Reply:
Done.